fn/mvar.go

```diff
@@ -0,0 +1,180 @@
+	current *atomic.Pointer[A]
+	// readers is used to wake all blocked readers when a new value is
+	// written.
+	readers chan chan A
+
+	// takers is used to wake a single taker when a new value is written.
+	takers chan chan A
+
+	// value is a bounded channel of size 1 that represents the core state
+	// oof the channel.
+	value chan A
+}
+
+// Zero initializes an MVar that has no values in it. In this state, TakeMVar
+// will block and PutMVar will immediately succeed.
+//
+// Zero : () -> MVar[A].
+func Zero[A any]() MVar[A] {
+	ptr := atomic.Pointer[A]{}
+
+	return MVar[A]{
+		current: &ptr,
+		readers: make(chan chan A),
+		takers:  make(chan chan A),
+		value:   make(chan A, 1),
+	}
+}
+
+// NewMVar initializes an MVar that has a value in it from the getgo. In this
+// state, TakeMVar will succeed immediately and PutMVar will block.
+//
+// NewMVar : A -> MVar[A].
+func NewMVar[A any](a A) MVar[A] {
+	z := Zero[A]()
+	z.value <- a
+	z.current.Store(&a)
+
+	return z
+}
+
+// Take will wait for a value to be put into the MVar and then immediately
+// take it out.
+//
+// Take : MVar[A] -> A.
+func (m *MVar[A]) Take() A {
+	select {
+	case v := <-m.value:
+		m.current.Store(nil)
+		return v
+	default:
+		t := make(chan A)
+		m.takers <- t
+		return <-t
+	}
+}
+
+// TryTake is the non-blocking version of TakeMVar, it will return an
+// None() Option if it would have blocked.
+//
+// TryTake : MVar[A] -> Option[A].
+func (m *MVar[A]) TryTake() Option[A] {
+	select {
+	case v := <-m.value:
+		m.current.Store(nil)
+		return Some(v)
+	default:
+		return None[A]()
+	}
+}
+
+// Put will wait for a value to be made empty and will immediately replace it
+// with the argument.
+//
+// Put : (MVar[A], A) -> ().
+func (m *MVar[A]) Put(a A) {
+readLoop:
+	// Give the newly put value to all of the waiting readers.
+	for {
+		select {
+		case r := <-m.readers:
+			r <- a
+		default:
+			break readLoop
+		}
+	}
+
+	// Give the newly put value to a single taker if one exists. If there
+	// are no available takers, then store it in the MVar. Since the value
+	// channel is bounded with capacity 1, subsequent put operations will
+	// block.
+	select {
+	case t := <-m.takers:
+		t <- a
+	default:
+		m.value <- a
+		m.current.Store(&a)
+	}
+}
+
+// TryPut is the non-blocking version of Put and will return true if the MVar is
+// successfully set.
+//
+// TryPut : (MVar[A], A) -> bool.
+func (m *MVar[A]) TryPut(a A) bool {
+	select {
+	case m.value <- a:
+		m.current.Store(&a)
+		return true
+	default:
+		return false
+	}
+}
+
+// Read will atomically read the contents of the MVar. If the MVar is empty,
+// Read will block until a value is put in. Callers of Read are guaranteed to
+// be woken up before callers of Take.
+//
+// Read : MVar[A] -> A.
+func (m *MVar[A]) Read() A {
+	// Check to see if MVar has something in it.
+	if ptr := m.current.Load(); ptr != nil {
+		return *ptr
+	}
+
+	// It's empty so we need to wait.
+	r := make(chan A)
```

Comment:
I think this might miss setting the current value.
1) We check that the current value is set => it's not yet set
2) We set the current value elsewhere => notify readers
3) we add a new reader
4) miss 2)

Alternatively we could use a conditional variable which is well tested to implement the same behavior.

Reply:
Good point. I can adjust this. I am avoiding using Cond's here as an earlier version of this code used them and it prevented my ability to guarantee that readers would all be woken before the takers without a bunch more machinery.

Reply:
I've fixed this using Conds.